WIP debug vs. release builds (#307)
memfault/interrupt: _drafts/debug-vs-release-builds.md

---
title: "Pocket article: Debug vs. Release Builds Considered Harmful"
# title: "Pocket article: Debug vs. Release Builds and why you shouldn't separate them"
description: "Why having separate Debug/Release builds should be avoided"
author: noah
image:
---

<!-- excerpt start -->

> Although having separate debug and release builds may appear to be a reasonable

Author: :+1: yeah, this intro needs big rework, it's not clear at all

than it solves. Maintaining two different builds is burdensome, and it can lead
to inconsistencies between the two, which can cause confusion and frustration
during development.

<!-- excerpt end -->

{% include newsletter.html %}

{% include toc.html %}

## Debug vs. Release Builds for Embedded Applications

Many embedded development IDEs provide separate "Debug" and "Release" build
configurations out of the box. Generally the differences between those can be
summarized as:

| Build config | Debug info enabled?          | Compiler optimization?      | `NDEBUG` preprocessor definition?          |
| ------------ | ---------------------------- | --------------------------- | ------------------------------------------ |
| **Debug**    | ✅ enabled (`-g` or similar) | low (`-O0` or `-Og`)        | `DEBUG` is set, or `#define NDEBUG 0`, etc |
| **Release**  | ❌ disabled                  | high (`-O3` or `-Os`/`-Oz`) | `NDEBUG` is set                            |

Let's go one-by-one through these items!

## Debug info

Debug info ("debugging information") is metadata that's included in the output
symbol file (typically `project.elf`/`project.out`/`project.axf`, but almost

> always it's an ELF file).

Reviewer: Might be good to say ELF is used regardless of the file's extension? I guess that's not technically correct with .axf I think?

Author: axf is elf too! wish these ide's had used .elf to start but too late now i suppose

  for some C++ projects, where the type information can grow dramatically (for
  example, building the Chrome browser with debug info enabled can [exceed
  4GiB](https://randomascii.wordpress.com/2023/03/08/when-debug-symbols-get-large/)
  ❗)

A quick worked example demonstrating that compiling with debug info has **zero**
impact on the final program:

```bash
# a simple hello world program
❯ echo > hello.c <<EOF
#include <stdio.h>
int main(void) {
  printf("hello!\n");
  return 0;
}
EOF

# compile with debug info disabled
❯ gcc -o hello-no-debug hello.c

# compile it again with debug info enabled
❯ gcc -ggdb3 -o hello-with-debug hello.c

# run the 'strip' tool to remove all data not part of the program contents.
# note that we're also removing the (non-loaded) gnu build ID section, because
# the gnu build ID is computed over the program contents AND debug information;
# 'strip' doesn't remove it by default.
# see https://interrupt.memfault.com/blog/reproducible-firmware-builds for
# details!
❯ strip --remove-section=.note.gnu.build-id hello-no-debug hello-with-debug

# compare the two files; they're bit-for-bit identical

> ❯ diff hello-no-debug hello-with-debug && echo 'identical!'

Reviewer 2: cool demo

information (significant data structures or variables, call stack) is present.
I've never hit a roadblock due to compiler optimization in step-debugging- worst
case I need to peek at a few assembly instructions to see what's going on, but
that's quite rare.

There's a HUGE downside to having different optimization levels for debug vs.
release builds: we end up with a **different** executable:

- sizes will likely be different: stack or heap usage (affecting program
  stability!), code space (affecting flash memory allocation, OTA)
- different machine code executed by the CPU: this can impact **timing** (often
  where the nastiest bugs show up) and actual **program validity**, eg
  assumptions you are making about certain spots in your program may no longer
  be valid with different optimization levels (the classic one is instruction
  reordering causing concurrent access bugs!)

Because your debug (testing) build is not running the same code, you may see
these bugs only show up during field test, rather than when the device is on
your desk and connected to a debugger!

It's also generally not desirable to adjust compiler optimization level when a

> project is nearing time to ship, to decrease risk- if the debug + release builds

Author: :sweat_smile: it's a terrible habit! need to break it

are using the same optimization settings, it's less likely to hit a surprise
out-of-space issue that can stop a project in its tracks 😱.

**Summary:**

- use the same optimization settings for all builds to decrease risk

## `NDEBUG` preprocessor definition

Often these Debug/Release builds will have some preprocessor defined symbols
such as:

- `DEBUG=1` (debug enabled)
- `NDEBUG=0` (indicating this is NOT a non-debug build)

Usually these flags will be used to compile in features like this to the Debug
builds:

- enable an external CLI for testing (this is so useful in production devices! just
  secure it properly)
- disable runtime asserts (assuming your system can tolerate a reset due to an
  `ASSERT` condition, these are stupendously valuable in production builds- see
  [this post]({% post_url 2019-11-20-cortex-m-hardfault-debug %}) for more

>   details)

Author: whoops yeah, this should be that link! tyty

- changing default log levels, i.e. enabling `INFO` or `DEBUG` levels- this can
  impact timing or memory/bandwidth usage, so generally it's preferable to stick

>   to one level and only increase in select cases when you need extra information

Reviewer 2: So the recommendation is to use a level of logging  in development that is acceptable in production? You might want to clarify this.

I am not sure that is always true. I think there is a lot of value in collecting extra logs in development and field testing, although you cannot afford to keep the extra logs in production because of flash wear. What would you say to this argument?

Author: very good point, I'll include it! I like the idea of runtime configuration here but not always practical, so this is a great example of "yes, production builds may want to be different"

  (eg enable it only for a single file actively being debugged)

I've seen the `DEBUG` flag used pretty aggressively to cut out entire subsystems
that aren't strictly required for production. This can cause a lot of confusion,
for example the following scenario:

1. production test build encounters a problem
2. engineer goes to retrieve logs, realizes the logging subsystem was entirely
   disabled in the "release" build
3. 😩

**Summary:**

> - it can be quite hazardous to allow splitting features based on preprocessor

Reviewer 2: so you are recommending keeping the features in and then doing configuration (feature flags) over preprocessor usage?

>   flags, and should only be done if there's a strict requirement. recommendation

Reviewer 2: ```suggestion
  flags, and should only be done if there's a strict requirement. Recommendation
```

---
title: "Pocket article: Debug vs. Release Builds Considered Harmful"
# title: "Pocket article: Debug vs. Release Builds and why you shouldn't separate them"
description: "Why having separate Debug/Release builds should be avoided"
author: noah
image:
---

<!-- excerpt start -->

Separate "debug" and "release" builds are very common in embedded development.

> Typically the notion is faster debug speed (less aggressive compiler

Reviewer: ```suggestion
Typically the notion is improved debug capabilities (less aggressive compiler
```

symbol file (typically `project.elf`/`project.out`/`project.axf`, but almost
always it's an ELF file).

This debugging information is placed into the symbol file in "non-loadable"
sections, which means:

- When flashing the symbol file to a device, the debug info is not loaded (it's
  completely ignored)
- When converting the symbol file to a `.bin` or `.hex` format for factory
  programming, the debug info is omitted

Since enabling debug info has no impact on the final loaded executable, the only
downsides to enabling it are:

- _Slightly_ longer build times (can be ~10%, but difficult to measure, and
  negligible if you're using [build caching]({% post_url
  2020-02-11-improving-compilation-times-c-cpp-projects %})!)
- Much larger symbol file size (can go from approximately the on-target size of
  the program, to many megabytes). This is generally not a problem, except for
  some C++ projects, where the type information can grow dramatically (for
  example, building the Chrome browser with debug info enabled can
  [exceed 4GiB](https://randomascii.wordpress.com/2023/03/08/when-debug-symbols-get-large/)
  ❗)

> A quick worked example demonstrating that compiling with debug info has **zero**

Reviewer: ```suggestion
A quick example demonstrating that compiling with debug info has **zero**
```

1. Changing **log verbosity** (disabling DEBUG or INFO levels in production, to
   prevent unnecessary flash wear)
   - I'd recommend instead, sticking with the **production** level, and in
     exceptional cases (problematic device), upgrade to a more verbose level
     temporarily
   - Depending on the particular system, it may be preferable to adjust
     verbosity at runtime, and keep the build the same. For example, leave
     WARNING and ERROR enabled by default, but allow increasing verbosity via a
     persistent on-device setting, or a temporary runtime flag.
   - Be especially wary of builds with different log verbosity- it can impact
     timing and memory usage _dramatically_!
2. Disabling an **"engineering interface"** (eg ssh or local serial console
   backdoor)
   - This can be a security requirement. An alternative is to build
     authentication into the engineering console, but this can be error prone so
     proceed cautiously.
   - Having the interface available (in an authenticated way) on production
     units can be extremely useful when diagnosing faulty customer units that
     have been returned for analysis!
3. Enabling **signed/encrypted builds**

>    - Also very tricky- you definitely want to be verifying the
>      signing/encryption implementation works throughout the development and test
>      cycle, instead of causing surprises at the end (execute-in-place encryption
>      too slow, key management not actually production ready, etc).

Reviewer: ```suggestion
   - Also very tricky! Best practice is to verify the
     signing/encryption implementation works throughout the development and test
     cycle, instead of causing surprises at the end (execute-in-place encryption
     too slow, key management not actually production ready, etc).
```

   - It might make the most sense to have a separate "development" key for

>      signing/encrypting development builds, entirely separate and with different

Reviewer: ```suggestion
     signing/encrypting development builds. A key that is separate and has different
```

     security concerns than the production key.

> 4. Disabling **runtime asserts** (assuming your system can tolerate a reset due
>    to an `ASSERT` condition, these are _stupendously_ valuable in production
>    builds- see [this
>    post]({% post_url 2019-11-05-asserts-in-embedded-systems %}) for more
>    details)

Reviewer: Just to match the other bullet points:

```suggestion
4. Disabling **runtime asserts**
    - If your system can tolerate a reset due
   to an `ASSERT` condition, these are _stupendously_ valuable in production
   builds- see [this
   post]({% post_url 2019-11-05-asserts-in-embedded-systems %}) for more
   details)
```

I've seen the `DEBUG` flag used pretty aggressively to cut out entire subsystems
that aren't strictly required for production. This can cause a lot of confusion,
for example the following scenario:

1. Production test build encounters a problem
2. Engineer goes to retrieve logs, realizes the logging subsystem was entirely
   disabled in the "release" build
3. 😩

**Summary:**

- It can be quite hazardous to allow splitting features based on preprocessor
  flags, and should only be done if there's a strict requirement. Recommendation
  is to only do it when generating builds from a CI server, if absolutely
  necessary.

I prefer feature-specific flags when necessary. This is less confusing than a
global `DEBUG` flag, and less likely to leak to other places:

```c
#if !defined(LOG_LEVEL)
#define LOG_LEVEL kLogLevel_Debug
#endif
```

## Summary

🙅🙅🙅🙅

Avoid separate "Debug" and "Release" builds at all costs!

> More seriously- there's a lot of downsides, and potentially wasted time, if a
> different build is used during development vs. production, and generally there's
> few cases where this is actually required!

Reviewer: ```suggestion
More seriously, there are many downsides(complexity, wasted time, etc) if a
different build is used during development vs. production. There are
few cases where this is actually required!
```

My advice is to consider very carefully if you have a need for a separate
Release build, and make sure you're ready to commit to supporting it.

Often it's sufficient to enable or disable a few small features via a
compilation flag for production builds. This can be done, for example, in your
CI server, when preparing a tag for release. That build can also be where
production keys/signing is performed, so it's all in one place and can be

> audited later. Avoid "Noah's laptop" builds making it to manufacturing 😅!

Reviewer: Amazing love it
